Sparksql select throws error after a select and insert sql

I add a testcase:

  @Test
  public void appendAndSelectData(TestInfo testInfo) {
    String datasetName = testInfo.getTestMethod().get().getName();
    String path = LanceConfig.getDatasetUri(dbPath.toString(), datasetName);
    spark.sql("CREATE OR REPLACE TABLE lance.`" + path + "` AS SELECT * FROM tmp_view");

    //    Dataset<Row> data = spark.sql("SELECT * FROM lance.`" + path + "`");
    //    assertEquals(2, data.count());
    spark.sql("INSERT INTO TABLE lance.`" + path + "` VALUES (3, 'Charlie')");
    Dataset<Row> data = spark.sql("SELECT * FROM lance.`" + path + "`");
    assertEquals(3, data.count());
    spark.sql("DROP TABLE lance.`" + path + "`");
  }
This test works fine. However, if we add back the commented-out code​​ as:

  @Test
  public void appendAndSelectData(TestInfo testInfo) {
    String datasetName = testInfo.getTestMethod().get().getName();
    String path = LanceConfig.getDatasetUri(dbPath.toString(), datasetName);
    spark.sql("CREATE OR REPLACE TABLE lance.`" + path + "` AS SELECT * FROM tmp_view");

    Dataset<Row> data = spark.sql("SELECT * FROM lance.`" + path + "`");
    assertEquals(2, data.count());
    spark.sql("INSERT INTO TABLE lance.`" + path + "` VALUES (3, 'Charlie')");
    data = spark.sql("SELECT * FROM lance.`" + path + "`");
    assertEquals(3, data.count());
    spark.sql("DROP TABLE lance.`" + path + "`");
  }
There will be an error at the seoncd select sql:

Driver stacktrace:
25/05/12 17:42:34 INFO DAGScheduler: ShuffleMapStage 5 (count at SparkConnectorWriteTest.java:247) failed in 0.030 s due to Job aborted due to stage failure: Task 2 in stage 5.0 failed 1 times, most recent failure: Lost task 2.0 in stage 5.0 (TID 7) (10.69.74.136 executor driver): java.lang.RuntimeException: java.lang.IndexOutOfBoundsException: Index: 2, Size: 2
at com.lancedb.lance.spark.internal.LanceFragmentScanner.create(LanceFragmentScanner.java:93)
at com.lancedb.lance.spark.internal.LanceDatasetAdapter.getFragmentScanner(LanceDatasetAdapter.java:91)
at com.lancedb.lance.spark.internal.LanceFragmentColumnarBatchScanner.create(LanceFragmentColumnarBatchScanner.java:39)
at com.lancedb.lance.spark.read.LanceColumnarPartitionReader.next(LanceColumnarPartitionReader.java:44)
at org.apache.spark.sql.execution.datasources.v2.PartitionIterator.hasNext(DataSourceRDD.scala:120)
at org.apache.spark.sql.execution.datasources.v2.MetricsIterator.hasNext(DataSourceRDD.scala:158)
at org.apache.spark.sql.execution.datasources.v2.DataSourceRDD$$anon$1.$anonfun$hasNext$1(DataSourceRDD.scala:63)
at org.apache.spark.sql.execution.datasources.v2.DataSourceRDD$$anon$1.$anonfun$hasNext$1$adapted(DataSourceRDD.scala:63)
at scala.Option.exists(Option.scala:376)
at org.apache.spark.sql.execution.datasources.v2.DataSourceRDD$$anon$1.hasNext(DataSourceRDD.scala:63)
at org.apache.spark.sql.execution.datasources.v2.DataSourceRDD$$anon$1.advanceToNextIter(DataSourceRDD.scala:97)
at org.apache.spark.sql.execution.datasources.v2.DataSourceRDD$$anon$1.hasNext(DataSourceRDD.scala:63)
at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.columnartorow_nextBatch_0$(Unknown Source)
at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.hashAgg_doAggregateWithoutKey_0$(Unknown Source)
at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
at org.apache.spark.sql.execution.WholeStageCodegenEvaluatorFactory$WholeStageCodegenPartitionEvaluator$$anon$1.hasNext(WholeStageCodegenEvaluatorFactory.scala:43)
at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:460)
at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:140)
at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:104)
at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:54)
at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:166)
at org.apache.spark.scheduler.Task.run(Task.scala:141)
at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:620)
at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:623)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.IndexOutOfBoundsException: Index: 2, Size: 2
at java.util.ArrayList.rangeCheck(ArrayList.java:659)
at java.util.ArrayList.get(ArrayList.java:435)
at com.lancedb.lance.spark.internal.LanceFragmentScanner.create(LanceFragmentScanner.java:66)
... 33 more

--- a/lance-spark-base_2.12/src/main/java/com/lancedb/lance/spark/internal/LanceFragmentScanner.java
+++ b/lance-spark-base_2.12/src/main/java/com/lancedb/lance/spark/internal/LanceFragmentScanner.java
@@ -34,35 +34,39 @@ import org.apache.spark.sql.types.StructType;
 import java.io.IOException;
 import java.util.Arrays;
 import java.util.List;
+import java.util.Map;
+import java.util.Objects;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 
 public class LanceFragmentScanner implements AutoCloseable {
-  private static LoadingCache<LanceConfig, List<Fragment>> LOADING_CACHE =
+  private static LoadingCache<CacheKey, Map<Integer, Fragment>> LOADING_CACHE =
       CacheBuilder.newBuilder()
           .maximumSize(100)
           .expireAfterAccess(1, TimeUnit.HOURS)
           .build(
-              new CacheLoader<LanceConfig, List<Fragment>>() {
+              new CacheLoader<CacheKey, Map<Integer, Fragment>>() {
                 @Override
-                public List<Fragment> load(LanceConfig config) throws Exception {
+                public Map<Integer, Fragment> load(CacheKey key) throws Exception {
                   BufferAllocator allocator = LanceDatasetAdapter.allocator;
+                  LanceConfig config = key.getConfig();
                   ReadOptions options = SparkOptions.genReadOptionFromConfig(config);
                   Dataset dataset = Dataset.open(allocator, config.getDatasetUri(), options);
-                  return dataset.getFragments();
+                  return dataset.getFragments().stream()
+                      .collect(Collectors.toMap(Fragment::getId, f -> f));
                 }
               });
-  private LanceScanner scanner;
+  private final LanceScanner scanner;
 
   private LanceFragmentScanner(LanceScanner scanner) {
     this.scanner = scanner;
   }
 
   public static LanceFragmentScanner create(int fragmentId, LanceInputPartition inputPartition) {
-    LanceScanner scanner = null;
     try {
       LanceConfig config = inputPartition.getConfig();
-      List<Fragment> cachedFragments = LOADING_CACHE.get(config);
+      CacheKey key = new CacheKey(config, inputPartition.getScanId());
+      Map<Integer, Fragment> cachedFragments = LOADING_CACHE.get(key);
       Fragment fragment = cachedFragments.get(fragmentId);
       ScanOptions.Builder scanOptions = new ScanOptions.Builder();
       scanOptions.columns(getColumnNames(inputPartition.getSchema()));
@@ -81,18 +85,10 @@ public class LanceFragmentScanner implements AutoCloseable {
       if (inputPartition.getTopNSortOrders().isPresent()) {
         scanOptions.setColumnOrderings(inputPartition.getTopNSortOrders().get());
       }
-      scanner = fragment.newScan(scanOptions.build());
-    } catch (Throwable t) {
-      if (scanner != null) {
-        try {
-          scanner.close();
-        } catch (Throwable it) {
-          t.addSuppressed(it);
-        }
-      }
-      throw new RuntimeException(t);
+      return new LanceFragmentScanner(fragment.newScan(scanOptions.build()));
+    } catch (Throwable throwable) {
+      throw new RuntimeException(throwable);
     }
-    return new LanceFragmentScanner(scanner);
   }
 
   /** @return the arrow reader. The caller is responsible for closing the reader */
@@ -129,5 +125,31 @@ public class LanceFragmentScanner implements AutoCloseable {
     return Arrays.stream(schema.fields())
         .map(StructField::name)
         .anyMatch(name -> name.equals(LanceConstant.ROW_ADDRESS));
+  }
+
+  private static class CacheKey {
+    private final LanceConfig config;
+    private final String scanId;
+
+    CacheKey(LanceConfig config, String scanId) {
+      this.config = config;
+      this.scanId = scanId;
+    }
+
+    public LanceConfig getConfig() {
+      return config;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == null || getClass() != o.getClass()) return false;
+      CacheKey cacheKey = (CacheKey) o;
+      return Objects.equals(config, cacheKey.config) && Objects.equals(scanId, cacheKey.scanId);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(config, scanId);
+    }
   }
 }

--- a/lance-spark-base_2.12/src/main/java/com/lancedb/lance/spark/read/LanceInputPartition.java
+++ b/lance-spark-base_2.12/src/main/java/com/lancedb/lance/spark/read/LanceInputPartition.java
@@ -33,22 +33,7 @@ public class LanceInputPartition implements InputPartition {
   private final Optional<Integer> limit;
   private final Optional<Integer> offset;
   private final Optional<List<ColumnOrdering>> topNSortOrders;
-
-  public LanceInputPartition(
-      StructType schema,
-      int partitionId,
-      LanceSplit lanceSplit,
-      LanceConfig config,
-      Optional<String> whereCondition) {
-    this.schema = schema;
-    this.partitionId = partitionId;
-    this.lanceSplit = lanceSplit;
-    this.config = config;
-    this.whereCondition = whereCondition;
-    this.limit = Optional.empty();
-    this.offset = Optional.empty();
-    this.topNSortOrders = Optional.empty();
-  }
+  private final String scanId;
 
   public LanceInputPartition(
       StructType schema,
@@ -56,16 +41,17 @@ public class LanceInputPartition implements InputPartition {
       LanceSplit lanceSplit,
       LanceConfig config,
       Optional<String> whereCondition,
-      Optional<Integer> limit,
-      Optional<Integer> offset) {
-    this.schema = schema;
-    this.partitionId = partitionId;
-    this.lanceSplit = lanceSplit;
-    this.config = config;
-    this.whereCondition = whereCondition;
-    this.limit = limit;
-    this.offset = offset;
-    this.topNSortOrders = Optional.empty();
+      String scanId) {
+    this(
+        schema,
+        partitionId,
+        lanceSplit,
+        config,
+        whereCondition,
+        Optional.empty(),
+        Optional.empty(),
+        Optional.empty(),
+        scanId);
   }
 
   public LanceInputPartition(
@@ -76,7 +62,8 @@ public class LanceInputPartition implements InputPartition {
       Optional<String> whereCondition,
       Optional<Integer> limit,
       Optional<Integer> offset,
-      Optional<List<ColumnOrdering>> topNSortOrders) {
+      Optional<List<ColumnOrdering>> topNSortOrders,
+      String scanId) {
     this.schema = schema;
     this.partitionId = partitionId;
     this.lanceSplit = lanceSplit;
@@ -85,6 +72,7 @@ public class LanceInputPartition implements InputPartition {
     this.limit = limit;
     this.offset = offset;
     this.topNSortOrders = topNSortOrders;
+    this.scanId = scanId;
   }
 
   public StructType getSchema() {
@@ -117,5 +105,9 @@ public class LanceInputPartition implements InputPartition {
 
   public Optional<List<ColumnOrdering>> getTopNSortOrders() {
     return topNSortOrders;
+  }
+
+  public String getScanId() {
+    return scanId;
   }
 }

--- a/lance-spark-base_2.12/src/main/java/com/lancedb/lance/spark/read/LanceScan.java
+++ b/lance-spark-base_2.12/src/main/java/com/lancedb/lance/spark/read/LanceScan.java
@@ -34,6 +34,7 @@ import scala.collection.mutable.HashMap;
 
 import java.io.Serializable;
 import java.util.List;
+import java.util.UUID;
 import java.util.stream.IntStream;
 
 public class LanceScan
@@ -46,6 +47,7 @@ public class LanceScan
   private final Optional<Integer> limit;
   private final Optional<Integer> offset;
   private final Optional<List<ColumnOrdering>> topNSortOrders;
+  private final String scanId = UUID.randomUUID().toString();
 
   public LanceScan(
       StructType schema,
@@ -81,7 +83,8 @@ public class LanceScan
                     whereConditions,
                     limit,
                     offset,
-                    topNSortOrders))
+                    topNSortOrders,
+                    scanId))
         .toArray(InputPartition[]::new);
   }
 

--- a/lance-spark-base_2.12/src/test/java/com/lancedb/lance/spark/TestUtils.java
+++ b/lance-spark-base_2.12/src/test/java/com/lancedb/lance/spark/TestUtils.java
@@ -72,7 +72,12 @@ public class TestUtils {
       lanceConfig = LanceConfig.from(datasetUri);
       inputPartition =
           new LanceInputPartition(
-              schema, 0, new LanceSplit(Arrays.asList(0, 1)), lanceConfig, Optional.empty());
+              schema,
+              0,
+              new LanceSplit(Arrays.asList(0, 1)),
+              lanceConfig,
+              Optional.empty(),
+              "test");
     }
   }
 }

--- a/lance-spark-base_2.12/src/test/java/com/lancedb/lance/spark/read/LanceColumnarPartitionReaderTest.java
+++ b/lance-spark-base_2.12/src/test/java/com/lancedb/lance/spark/read/LanceColumnarPartitionReaderTest.java
@@ -37,7 +37,8 @@ public class LanceColumnarPartitionReaderTest {
             0,
             split,
             TestUtils.TestTable1Config.lanceConfig,
-            Optional.empty());
+            Optional.empty(),
+            "test");
     try (LanceColumnarPartitionReader reader = new LanceColumnarPartitionReader(partition)) {
       List<List<Long>> expectedValues = TestUtils.TestTable1Config.expectedValues;
       int rowIndex = 0;
@@ -73,7 +74,9 @@ public class LanceColumnarPartitionReaderTest {
             TestUtils.TestTable1Config.lanceConfig,
             Optional.empty(),
             Optional.of(1),
-            Optional.of(1));
+            Optional.of(1),
+            Optional.empty(),
+            "testOffsetAndLimit");
     try (LanceColumnarPartitionReader reader = new LanceColumnarPartitionReader(partition)) {
       List<List<Long>> expectedValues = TestUtils.TestTable1Config.expectedValues;
       int rowIndex = 1;
@@ -112,7 +115,8 @@ public class LanceColumnarPartitionReaderTest {
             Optional.empty(),
             Optional.of(1),
             Optional.empty(),
-            Optional.of(Collections.singletonList(builder.build())));
+            Optional.of(Collections.singletonList(builder.build())),
+            "testTopN");
     try (LanceColumnarPartitionReader reader = new LanceColumnarPartitionReader(partition)) {
       List<List<Long>> expectedValues = TestUtils.TestTable1Config.expectedValues;
 

--- a/lance-spark-base_2.12/src/test/java/com/lancedb/lance/spark/read/LanceDatasetReadTest.java
+++ b/lance-spark-base_2.12/src/test/java/com/lancedb/lance/spark/read/LanceDatasetReadTest.java
@@ -92,7 +92,8 @@ public class LanceDatasetReadTest {
                 0,
                 new LanceSplit(Arrays.asList(fragment)),
                 TestUtils.TestTable1Config.lanceConfig,
-                Optional.empty()))) {
+                Optional.empty(),
+                "validateFragment"))) {
       try (ArrowReader reader = scanner.getArrowReader()) {
         VectorSchemaRoot root = reader.getVectorSchemaRoot();
         assertNotNull(root);

